### PR TITLE
Warn if log groups are used as log level, or vice versa -- SIMICS-21138

### PR DESCRIPTION
**Pull request by lwaern-intel**
Depends on #186 

The verification of this is gonna be interesting. It will complain about what https://github.com/intel-innersource/applications.simulators.simics.simics-base/pull/4184 addresses, but what else?

### CODE REVIEW

**Comment by lwaern-intel** [comment]
Oh there's already plenty... yeah we should have implemented these sooner.

**Comment by syssimics** [comment]
Verification [#11785943](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=11785943): fail

**Comment by lwaern-intel** [comment]
The warnings are now only enabled by default on Simics 7.

I will run a VP pretest against this to check if the `as_int` use reveals an existing bug like it did for `trainings`.

**Comment by syssimics** [comment]
Verification [#12157070](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=12157070): fail

File: RELEASENOTES.docu [hunk 1]
```diff
@@ -1724,4 +1724,11 @@ extern typedef struct { } my_type_t;</pre> </add-note></build-id>
   <build-id value="next"><add-note> Added an error if a device declares
       more than 63 log groups (including the two built-in log groups.)
   </add-note></build-id>
+  <build-id value="next"><add-note> Added a pair of warnings for if the
+      specified log level of a log statement is likely intended to instead
+      specify the log groups (<tt>WLGROUPSASLEV</tt>), or vice versa
+      (<tt>WLLEVASGROUPS</tt>) <bug number="22018374443"/>.
+      These warnings will only be enabled by default in Simics 7. In Simics 6
+      and below they must be explicitly enabled through the use of the
+      <tt>--warn</tt> flag. </add-note></build-id>
```
Comment (mandolaerik) [inline RELEASENOTES.docu:1733]: We can consider merging the warnings to one? I doubt that anyone ever will only want one of them, and easier for the user to pass a single warning flag (and easier for us to explain it: "..through the use of the <tt>--warn=WLOGGROUP</tt> flag"). We can still parameterize the warning to give descriptive messages in both cases.

Comment (mandolaerik) [reply RELEASENOTES.docu:1733]: Technically, it's enabled by default with SIMICS_API=7. But your wording is sufficient for a releasenote.

File: py/dml/messages.py [hunk 2]
```diff
@@ -1895,6 +1895,44 @@ def __init__(self, site, old, new, maybe_intended):
                       if maybe_intended else '')
         DMLWarning.__init__(self, site, old, new, suggestion)
 
+class WLGROUPSASLEV(DMLWarning):
+    """
+    The specified log level of a `log` looks as though you meant to specify the
+    log groups instead. For example:
+    ```
+    log info, some_log_group: ...;
+    log info, (some_log_group | another_log_group), 1: ...;
```
Comment (mandolaerik) [inline py/dml/messages.py:1904]: Note that we generally discourage `log info` on level 1 in production code (it's good for printf debugging only). So better to use e.g. `spec_viol` for the level 1 cases.

File: py/dml/messages.py [hunk 3]
```diff
@@ -1895,6 +1895,44 @@ def __init__(self, site, old, new, maybe_intended):
...
+    explicitly specify the log level, and specify it first.
+    ```
+    log info, 1, some_log_group: ...;
+    log info, 1, (some_log_group | another_log_group): ...;
+    ```
+    """
+    fmt = ("the specified log level is likely intended to instead be a log "
+           + "groups specification. "
+           + "In order to specify log groups, the log level must be "
+           + "(explicitly) specified beforehand.")
```
Comment (mandolaerik) [inline py/dml/messages.py:1916]: You should perhaps mention "disabled by default when SIMICS_API <= 6"

File: py/dml/dmlc.py [hunk 4]
```diff
@@ -481,6 +481,26 @@ def show_illegal_attributes(option, opt, value, parser):
...
+              " or older")
+        sys.exit(1)
+
+    # This pair of warnings are disabled by default below Simics 7 due to sheer
+    # prominence of the issues they warn about in existing code.
+    # By only enabling the warnings in Simics 7, we allow existing codebases
+    # to handle the bugs as part of migration, instead of suddenly
+    # overwhelming them with a truly massive amount of warnings in an
+    # intermediary release.
+    if options.api_version < '7':
```
Comment (mandolaerik) [inline py/dml/dmlc.py:500]: I prefer to phrase it as `not in {'4.8', '5', '6'}`; both for subtle subjective reasons and because `'10'` < `'7'`.

Comment (lwaern-intel) [reply py/dml/dmlc.py:500]: Oh yeah, true. This is not the only place in DMLC we do checks like this -- we should do a once-over.

File: py/dml/messages.py (hunk 3 above)
Comment (lwaern-intel) [reply py/dml/messages.py:1916]: In the docstring, that is? Having it be in the error message makes no sense.

File: RELEASENOTES.docu (hunk 1 above)
Comment (lwaern-intel) [reply RELEASENOTES.docu:1733]: True. How about `WLLEVGROUPSMIXUP` as the unified warning? Or maybe just `WLOGMIXUP`?

File: py/dml/messages.py [hunk 5]
```diff
@@ -1895,43 +1895,79 @@ def __init__(self, site, old, new, maybe_intended):
...
-    fmt = ("the specified log groups are likely intended to instead be a log "
-           + "level specification. "
-           + "If so, reorder the specifications: the log level must be "
-           + "specified before the log groups.")
+
+    This warning is only enabled by default with Simics API version 7 or above.
+    """
+    fmt = ("log statement with likely misspecified log level(s) and log "
+           + "groups: %s")
+    def __init__(self, site, kind, level, later_level, groups):
```
Comment (lwaern-intel) [inline py/dml/messages.py:1927]: What a mess. Chime in if you think this is too complex.

File: test/tests.py [hunk 6]
```diff
@@ -1070,30 +1070,22 @@ def run_cc(self, cc_extraargs):
...
-        join(testdir, "1.4", "errors", "WLLEVASGROUPS.dml"),
-        dmlc_extraargs = ["--warn=WLLEVASGROUPS"]))
+        ["1.4", "errors", "WLOGMIXUP"],
+        join(testdir, "1.4", "errors", "WLOGMIXUP.dml"),
+        dmlc_extraargs = ["--warn=WLOGMIXUP"]))
 
 if get_simics_major() == "7":
     all_tests.append(CTestCase(
-        ["1.4", "errors", "T_ETYPE_integer_t"],
+        ["1.4", "errors", "ETYPE_integer_t"],
```
Comment (lwaern-intel) [inline test/tests.py:1083]: The `T_` prefix in subtest names are just redundant.

**Comment by syssimics** [comment]
Verification [#12174402](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=12174402): fail

**Comment by syssimics** [comment]
Verification [#12211754](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=12211754): fail

**Comment by syssimics** [comment]
Verification [#12221962](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=12221962): fail

**Comment by syssimics** [comment]
Verification [#12222398](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=12222398): pass

**Review by mandolaerik — APPROVED** [review]

File: test/tests.py [hunk 7]
```diff
@@ -1070,15 +1070,23 @@ def run_cc(self, cc_extraargs):
...
         join(testdir, "1.4", "errors", "ETYPE_integer_t.dml"),
         api_version="7"))
+    all_tests.append(CTestCase(
+        ["1.4", "errors", "WLOGMIXUP"],
+        join(testdir, "1.4", "errors", "WLOGMIXUP.dml"),
+        api_version="7"))
```
Comment (mandolaerik) [inline test/tests.py:1089]: This is fine, but one alternative to consider is `/// DMLC-FLAG --api-version=7` in the test. Deserves an explanatory comment too ("only enabled by default in 7 and newer"): in a couple of years, api=7 will indicate it tests some legacy behaviour.

Comment (lwaern-intel) [reply test/tests.py:1089]: `--api-version=7` won't be accepted in simics-base/6, so that's not an option.

**Pull request merged**